### ISSUE TITLE
[Help]: GMT command to mask phasefilt with a mask in radar coordinates

### ISSUE THREAD

**Issue by SteffanDavies** [gmtsar/gmtsar#612]
I am trying a new approach to see if I can improve RAM utilization and processing time for extensive AOIs.

I have a mask in radar coordinates that I want to use to convert all values outside of the mask in the phasefilt.grd file to NaN. The idea is to improve unwrapping speed by unwrapping only in the area inside the mask, while retaining the original raster size of the wrapped phase.

phasefilt.grd:

![image](https://user-images.githubusercontent.com/50405969/210035019-85c48ab4-4539-42f3-b052-6271ef34b52d.png)

mask.grd:

![image](https://user-images.githubusercontent.com/50405969/210035035-3675bf48-bbf3-4a10-99ed-bb4a436315fe.png)

So considering I am not familiar with GMT, which command can I use to do the following:

Keeping the original grid size of phasefilt.grd, keep in phasefilt.grd what is 1 in mask.grd, and NaN what is not 1 in mask.grd.

Then I can apply the mask to all the phasefilt.grd files and run parallel unwrapping to see if there is any improvement and compare results to unwrapping the entire original scene.

**Comment by Xiaohua-Eric-Xu** [comment]
Hi @SteffanDavies, you could 
1) draw your polygon (of the mask) on Google Earth or any GIS software and output to a text file.
2) use `gmt grdtrack your_table.txt -Gdem.grd > your_table_topo.txt` to get their elevation from your dem.grd
3) run `SAT_llt2rat` to figure out the polygon positions in radar coordinates.
4) run `gmt grdmask your_table_in_RA.txt -Rphasefilt.grd -Gyour_mask.grd -Nout/edge/in` to create a mask in radar coordinates. Here out, edge and in denotes values for pixels outside, on the path, or inside the polygon.

**Comment by SteffanDavies** [comment]
> 

I'm trying to figure out how to output my Geopackage to a text file. QGIS shows no options of exporting as txt (although I can export as a .csv with WKT format for a polygon) and neither does Google Earth.

**Comment by Xiaohua-Eric-Xu** [comment]
> I'm trying to figure out how to output my Geopackage to a text file. QGIS shows no options of exporting as txt (although I can export as a .csv with WKT format for a polygon) and neither does Google Earth. I'm not sure either of these solutions export to a GMT compatible format by default.
> 
> Edit: I have managed by using GE and exporting to .kml and then using `gmt kml2gmt` command. However I need to look into a more automated solution. QGIS outputs to kml but it's not working using the gmt command. Maybe it's an option during QGIS export than need to change.

I usually just output the polygon with GE as .kml file and then run `gmt kml2gmt` to convert things to a text file. Not sure how to do the exact same thing  with QGIS. Maybe you need some special arguments like `-Fp` to tell `gmt` it is a polygon? I am not sure.

**Comment by SteffanDavies** [comment]
I am not having much success with SAT_llt2rat.

Note that I am working on the merged products before unwrapping.

I have a 3 column table with X Y and DEM (Z) in Geographic projection:

```
27.23877198     38.4189063138   105.347946097
27.236439817    38.4192282648   94.7131959102
27.2361581289   38.4192813498   92.3102257023
```
and so on...

I have a copy of the supermaster.PRM
I have a copy of the LED file contained in the supermaster.PRM

I run `SAT_llt2rat ./supermaster.PRM 1 < kmltable_dem.txt > kmltable_dem_ra.txt` and nothing happens.

**Comment by Xiaohua-Eric-Xu** [comment]
Here's what I got with a random PRM file I got. Not sure what happened. Nothing in kmltable_dem_ra.txt?
```
[IG-A49515:~/Downloads/temp] xiaohua% cat kmltable_dem.txt 
27.23877198     38.4189063138   105.347946097
27.236439817    38.4192282648   94.7131959102
27.2361581289   38.4192813498   92.3102257023
[IG-A49515:~/Downloads/temp] xiaohua% SAT_llt2rat S1_20200402_033537_F3.PRM 1 < kmltable_dem.txt > kmltable_dem_ra.txt
[IG-A49515:~/Downloads/temp] xiaohua% cat kmltable_dem_ra.txt 
305706.044848240 -31912.363464460 -2698.037524505 27.238771980 38.418906314 
305790.837408462 -31914.017561318 -2708.788998640 27.236439817 38.419228265 
305801.281125136 -31914.337606278 -2711.211211477 27.236158129 38.419281350 
[IG-A49515:~/Downloads/temp] xiaohua% 
```

**Comment by SteffanDavies** [comment]
I'm close to having a solution using python, the problem is trans.dat is converting my mask.grd into radar in a 32/64 resolution instead of the 4/1 resolution of my phasefilt.grd files.

If I had an equal 4 to 1 grid this would be much easier.

Could explain why it's set to this resolution?

**Comment by SteffanDavies** [comment]
> Here's what I got with a random PRM file I got. Not sure what happened. Nothing in kmltable_dem_ra.txt?
> 
> ```
> [IG-A49515:~/Downloads/temp] xiaohua% cat kmltable_dem.txt 
> 27.23877198     38.4189063138   105.347946097
> 27.236439817    38.4192282648   94.7131959102
> 27.2361581289   38.4192813498   92.3102257023
> [IG-A49515:~/Downloads/temp] xiaohua% SAT_llt2rat S1_20200402_033537_F3.PRM 1 < kmltable_dem.txt > kmltable_dem_ra.txt
> [IG-A49515:~/Downloads/temp] xiaohua% cat kmltable_dem_ra.txt 
> 305706.044848240 -31912.363464460 -2698.037524505 27.238771980 38.418906314 
> 305790.837408462 -31914.017561318 -2708.788998640 27.236439817 38.419228265 
> 305801.281125136 -31914.337606278 -2711.211211477 27.236158129 38.419281350 
> [IG-A49515:~/Downloads/temp] xiaohua% 
> ```

I can provide you with my files because I have no idea why it's not working... empty result.

**Comment by Xiaohua-Eric-Xu** [comment]
Yes, that's the problem. @PaulWessel Does GMT has an option to not output the `-L <path name>` associated with the segment?

**Comment by PaulWessel** [comment]
That string is placed in the segment header, which is a normal part of a valid GMT table definition.  If some program cannot handle GMT segment headers then probably simplest to use grep or similar to dumb it down to a plain text file, e.g,

`gmt kml2gmt your file.kml | grep -v '^>' | thenextproess.`

Help me understand the context here: You have a KML file.  What geometry (points, lines, etc)? I am guessing you have one entity and **kml2gmt** writes output with a single segment header with the -L name?

**Comment by Xiaohua-Eric-Xu** [comment]
I see. It should be a polygon so we need to use `gmt kml2gmt kmltable_dem.kml -Fp | grep -v '^>' > kmltable_dem.gmt`.

**Comment by SteffanDavies** [comment]
@Xiaohua-Eric-Xu I have successfully applied a NaN mask to phasefilt but according to @mobigroup snaphu does not deal with NaNs. Should I mask the phasefilt with NaNs or 0s? 

Note that I can also mask the corr_mask outside study area AOI instead of simply using a coherence threshold which includes areas of sufficient coherence but no value to study area. 

The objective is to minimize RAM and decrease processing time.

**Comment by Xiaohua-Eric-Xu** [comment]
I think you could apply NaN or zero as NaNs are converted to zeros with `gmt grd2xyz -do0` in snaphu[_interp].csh. This type of masking will reduce the number of residues thus will speed up the solution from snaphu but not necessarily reduce memory cost unless you cut the images. You may also try other unwrappers that deal with sparse data.

**Comment by SteffanDavies** [comment]
> I think you could apply NaN or zero as NaNs are converted to zeros with `gmt grd2xyz -do0` in snaphu[_interp].csh. This type of masking will reduce the number of residues thus will speed up the solution from snaphu but not necessarily reduce memory cost unless you cut the images. You may also try other unwrappers that deal with sparse data.

In most situations where linear Infrastructure goes in SW-NE or NW-SE directions cutting is impossible, as it goes diagonally. In these cases, unwrapping in parallel using 16 cores can consume upwards of 200 GB ram for merged subswaths. 

I will run some tests this week as I expect much faster unwrapping with this masking however RAM continues to be a reason of concern for scalability.

**Comment by SteffanDavies** [comment]
> @SteffanDavies What exactly your problem? As discussed before SNAPHU can unwrap two merged scenes (full resolution) on 16 GB RAM host with 1 GB swap.

I was testing this approach without tiling. Seems to make almost no difference. Actually it's much worse due to nearers_grid.c taking hours due to the area being mostly NaN.

**Comment by SteffanDavies** [comment]
@Xiaohua-Eric-Xu Has GMTSAR implemented an alternative to nearest_grid.c for situations with large areas of no data?

An issue from last year was closed involving exactly this issue, but was closed.
https://github.com/gmtsar/gmtsar/issues/198

Any updates?

**Comment by mobigroup** [comment]
@SteffanDavies There was the solution (Python script) in PyGMTSAR sources (recently deleted but the old version is ok while for now it’s useless for PyGMTSAR). You would just replace the utility.

**Comment by mobigroup** [comment]
@SteffanDavies See https://github.com/mobigroup/gmtsar/blob/68dd53d335fe2adf58390c491ab8af7549ddd387/gmtsar/py/nearest_grid.py

PyGMTSAR is based on Python classes instead of the console scripts. I used the script for my SBAS shell scripts pipeline when I was checking new ideas, see https://github.com/mobigroup/gmtsar/tree/68dd53d335fe2adf58390c491ab8af7549ddd387/gmtsar/sh

**Comment by Xiaohua-Eric-Xu** [comment]
> @Xiaohua-Eric-Xu Has GMTSAR implemented an alternative to nearest_grid.c for situations with large areas of no data?
> 
> An issue from last year was closed involving exactly this problem. #198
> 
> Any updates?

No we haven't. For some cases (i.e. large area with no data), Matlab griddata with 'nearest' option runs faster. I am sure Python has similar modules.

**Comment by SteffanDavies** [comment]
I'm closing because nearest_grid is destroying any potential gains from masking in this case.
